File: backend/src/main/java/com/ufc/backend/backend/model/Student.java
package com.ufc.backend.backend.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.ufc.backend.backend.model.based.PersonBased;
import lombok.*;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Document
@ToString
public class Student extends PersonBased {
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    private String password;
    private String course;
    private List<Semester> semester = new ArrayList<>();
    @JsonIgnore
    private Set<Classes> notTakenClasses = new HashSet<>();

    public void addEmptySemester() {
        int index = this.semester.isEmpty() ? 1 : this.getSemester().size() + 1;
        this.getSemester().add(new Semester(index, null));
    }

    public void updateSemester(Semester newSemester) {
        Semester oldSemester = this.getSemester().get(newSemester.getIndex() - 1);
        oldSemester.getClasses().addAll(newSemester.getClasses());

        notifyTakenClasses(newSemester.getClasses());
        notifyNotTakenClasses(newSemester.getClasses());
    }

    private void notifyTakenClasses(Set<Classes> classes) {
        getClasses().addAll(classes);
    }

    private void notifyNotTakenClasses(Set<Classes> classes) {
        getNotTakenClasses().removeAll(classes);
    }

    @Override
    public boolean equals(Object obj) {
        if (getClass() != obj.getClass())
            return false;

        return this.getId().equals(((Student) obj).getId());
    }
}

File: backend/src/main/java/com/ufc/backend/backend/model/Student.kt
package com.ufc.backend.backend.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.ufc.backend.backend.commons.model.FeedbackBased
import com.ufc.backend.backend.commons.model.PersonBased
import com.ufc.backend.backend.exceptions.SemesterOutOfBoundsException
import com.ufc.backend.backend.commons.model.Person
import com.ufc.backend.backend.model.subject.Subject
import org.hibernate.annotations.GenericGenerator
import java.util.*
import javax.persistence.*

@Entity
class Student
    (
    @Id
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "org.hibernate.id.UUIDGenerator")
    val id: String,
    @Embedded
    override val person: Person = Person(),
    @Column
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    var password: String,
    @JsonIgnore
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "course_id", referencedColumnName = "id")
    var course: Course,
    @OneToMany(mappedBy = "student", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    var semester: MutableList<Semester>? = null,
    @JsonIgnore
    @ManyToMany
    @JoinTable(
        name = "students_subjects",
        joinColumns = [JoinColumn(name = "student_id")],
        inverseJoinColumns = [JoinColumn(name = "subject_id")]
    )
    var subjects: MutableSet<Subject>? = null
) : PersonBased, FeedbackBased() {

    private val newSemesterIndex = semester?.size?.minus(1) ?: 1

    fun getSemester(index: Int): Semester = semester?.get(index) ?: throw SemesterOutOfBoundsException()

    fun addEmptySemester() {
        semester?.plus(Semester(UUID.randomUUID().toString(), newSemesterIndex, null, this))
    }

    fun updateSemester(newSemester: Semester) {
        val oldSemester = getSemester(newSemester.semesterIndex)
        newSemester.subjects?.let { oldSemester.subjects?.addAll(it) }
        notifyTakenClasses(newSemester.subjects)
    }

    private fun notifyTakenClasses(subject: Collection<Subject>?) {
        if (subject != null) {
            subjects?.addAll(subject)
        }
    }

    override fun equals(other: Any?): Boolean {
        return if (javaClass != other!!.javaClass)
            false
        else id == (other as Student?)?.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }
}